src/url.cpp

```diff
@@ -457,4 +457,54 @@ namespace Url
         str.resize(dest);
     }
 
+    Url& Url::deparam(const std::unordered_set<std::string>& blacklist)
+    {
+        remove_params(params_, blacklist, ';');
+        remove_params(query_, blacklist, '&');
+        return *this;
+    }
+
+    void Url::remove_params(std::string& str, const std::unordered_set<std::string>& blacklist, const char separator)
+    {
+
+        std::vector<std::string> pieces;
+        size_t previous = 0;
+        for (size_t index = str.find(separator)
+            ; index != std::string::npos
+            ; previous = index + 1, index = str.find(separator, previous))
+        {
+            std::string piece = str.substr(previous, index - previous);
+            std::string name = piece.substr(0, piece.find('='));
+            std::transform(name.begin(), name.end(), name.begin(), ::tolower);
+
+            if (blacklist.find(name) == blacklist.end())
+            {
+                pieces.push_back(piece);
+            }
+        }
+
+        if (previous < str.length())
+        {
+            std::string piece = str.substr(previous);
+            std::string name = piece.substr(0, piece.find('='));
+            std::transform(name.begin(), name.end(), name.begin(), ::tolower);
```

Comment:
That's my understanding, too, and I suspect that it's the right / acceptable thing to do in this case. Any parameters we want to remove that aren't in ASCII will have to be escaped anyway, so it's not clear what to do in that case anyway -- when we want to replace `bücher` do we replace `b%C3%9Ccher` (bÜcher) _and_ `b%C3%BCcher` (bücher)? Or would we expect the argument to `deparam` to be the encoded version?

And then WRT hostnames (another spot where we use `::tolower`), I'm earnestly not sure if `bücher.com` and `bÜcher.com` are the same name -- they definitely punycode differently.

Thoughts?